File: src/main/java/june_2025/CloneGraph.java
package june_2025;

import java.util.*;

public class CloneGraph {

    public static void main(String[] args) {
        Node node1 = new Node(1);
        Node node2 = new Node(2);
        Node node3 = new Node(3);
        Node node4 = new Node(4);
        node1.neighbors.add(node2);
        node1.neighbors.add(node4);
        node2.neighbors.add(node1);
        node2.neighbors.add(node3);
        node3.neighbors.add(node2);
        node3.neighbors.add(node4);
        node4.neighbors.add(node1);
        node4.neighbors.add(node3);

        CloneGraph cloneGraph = new CloneGraph();
        Node node = cloneGraph.cloneGraph(node1);
        System.out.println(node);
    }

    // O(V + E) time | O(V) space
    public Node cloneGraph(Node node) {
        if (node == null) {
            return null;
        }
        Map<Integer, Node> adj = new HashMap<>();

        Queue<Node> queue = new LinkedList<>();
        queue.add(node);
        Set<Integer> visited = new HashSet<>();

        while (!queue.isEmpty()) {
            Node curr = queue.poll();
            int currVal = curr.val;
            if (visited.contains(currVal)) {
                continue;
            }
            visited.add(currVal);

            if (!adj.containsKey(currVal)) {
                Node newCurr = new Node(currVal);
                adj.put(newCurr.val, newCurr);
            }

            List<Node> neighbors = curr.neighbors;
            for (Node nei : neighbors) {
                queue.add(nei);
                int neiVal = nei.val;
                if (visited.contains(neiVal)) {
                    continue;
                }
                if (!adj.containsKey(neiVal)) {
                    Node newNeighbor = new Node(neiVal);
                    adj.put(newNeighbor.val, newNeighbor);
                }
                Node newCurr = adj.get(currVal);
                Node newNeighbor = adj.get(neiVal);
                newCurr.neighbors.add(newNeighbor);
                newNeighbor.neighbors.add(newCurr);
            }
        }
        return adj.get(node.val);
    }

    static class Node {
        public int val;
        public List<Node> neighbors;

        public Node(int val) {
            this.val = val;
            this.neighbors = new ArrayList<>();
        }
    }

}

File: python/clone_graph.py
class Solution(object):
    # O(v + e) time | O(v) space
    def cloneGraph(self, node):
        """
        :type node: Node
        :rtype: Node
        """
        if not node:
            return None
        
        origToNew = {}

        queue = []
        queue.append(node)
        visited = set()

        while queue:
            curr = queue.pop()
            if curr not in origToNew:
                origToNew[curr] = Node(curr.val)
            if curr.val in visited:
                continue
            visited.add(curr.val)

            newCurr = origToNew[curr]
            for nei in curr.neighbors:
                if nei not in origToNew:
                    origToNew[nei] = Node(nei.val)
                newCurr.neighbors.append(origToNew[nei])
                queue.append(nei)
        
        return origToNew[node]


class Node(object):
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []

node1 = Node(1)
node2 = Node(2)
node3 = Node(3)
node4 = Node(4)

node1.neighbors.append(node2)
node1.neighbors.append(node4)
node2.neighbors.append(node1)
node2.neighbors.append(node3)
node3.neighbors.append(node2)
node3.neighbors.append(node4)
node4.neighbors.append(node3)
node4.neighbors.append(node1)

solution = Solution()
print(solution.cloneGraph(node1))